pallets/mining/src/lib.rs

```diff
@@ -0,0 +1,376 @@
+        T::BitMiningResourceId::get()
+    }
+
+    pub fn is_mining_origin(who: &T::AccountId) -> bool {
+        let minting_origin = Self::minting_origin(who);
+        minting_origin == Some(())
+    }
+
+    pub fn ensure_admin(o: T::Origin) -> DispatchResult {
+        T::AdminOrigin::try_origin(o)
+            .map(|_| ())
+            .or_else(ensure_root)?;
+        Ok(())
+    }
+
+    fn do_mint(
+        who: T::AccountId,
+        amount: Balance,
+    ) -> DispatchResult {
+        if amount.is_zero() {
+            return Ok(());
+        }
+
+        ensure!(
+            Self::is_mining_origin(&who),
+            Error::<T>::NoPermission
+        );
+
+        let mining_treasury = Self::bit_mining_resource_account_id();
+        //Deposit Bit mining to mining treasury
+        T::MiningCurrency::deposit(Self::bit_mining_resource_currency_id(), &mining_treasury, amount)?;
+
+        Self::deposit_event(Event::<T>::MiningResourceMinted(amount));
+
+        Ok(())
+    }
+
+    fn do_burn(
+        who: T::AccountId,
+        amount: Balance,
+    ) -> DispatchResult {
+        if amount.is_zero() {
+            return Ok(());
+        }
+        ensure!(
+            Self::is_mining_origin(&who),
+            Error::<T>::NoPermission
+        );
+
+        let mining_treasury = Self::bit_mining_resource_account_id();
+        ensure!(
+            T::MiningCurrency::can_slash(Self::bit_mining_resource_currency_id(), &mining_treasury, amount),
+            Error::<T>::BalanceZero
+        );
+        //Deposit Bit mining to mining treasury
+        T::MiningCurrency::slash(Self::bit_mining_resource_currency_id(), &mining_treasury, amount);
+
+        Self::deposit_event(Event::<T>::MiningResourceBurned(amount));
+
+        Ok(())
+    }
+
+    fn do_deposit(
+        who: T::AccountId,
+        from: T::AccountId,
+        amount: Balance,
+    ) -> DispatchResult {
+        if amount.is_zero() {
+            return Ok(());
+        }
+
+        ensure!(
+            Self::is_mining_origin(&who),
+            Error::<T>::NoPermission
+        );
+
+        let mining_treasury = Self::bit_mining_resource_account_id();
+        ensure!(
+            T::MiningCurrency::free_balance(Self::bit_mining_resource_currency_id(), &from) >= amount,
+            Error::<T>::BalanceLow
+        );
+
+        T::MiningCurrency::transfer(Self::bit_mining_resource_currency_id(), &from, &mining_treasury, amount)?;
+
+        Self::deposit_event(Event::DepositMiningResource(from, amount.clone()));
+
+        Ok(())
+    }
+
+    fn do_withdraw(
+        who: T::AccountId,
+        dest: T::AccountId,
+        amount: Balance,
+    ) -> DispatchResult {
+        if amount.is_zero() || who == dest {
+            return Ok(());
+        }
+
+        ensure!(
+            Self::is_mining_origin(&who),
+            Error::<T>::NoPermission);
+
+        let mining_treasury = Self::bit_mining_resource_account_id();
+        ensure!(
```

Comment:
When withdraw bit to blockchain, I think we should check the balance in the centralized place. i.e. bit country database

Reply:
All extrinsic calls from these method should be validated on offchain process before triggering these calls.